src/test/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounderTest.java

```diff
@@ -97,14 +112,228 @@ public void groundConstraintAlreadyGround() {
+		Program program = PARSER.parse("a(1). "
+				+ "c(X) :- a(X), b(X). "
+				+ "b(X) :- something(X).");
+		testLaxGrounderHeuristicTolerance(program, 0, literal("a", "X"), 1, 1, 0, true, Arrays.asList(1));
+	}
+	
+	@Test
+	public void testLaxGrounderHeuristicTolerance_1_reject() {
+		Program program = PARSER.parse("a(1). "
+				+ "c(X) :- a(X), b(X), b(X+1). "
+				+ "b(X) :- something(X).");
+		testLaxGrounderHeuristicTolerance(program, 0, literal("a", "X"), 1, 1, 0, false, Arrays.asList(2));
+	}
+
+	@Test
+	public void testLaxGrounderHeuristicTolerance_2_accept() {
+		Program program = PARSER.parse("a(1). "
+				+ "c(X) :- a(X), b(X), b(X+1). "
+				+ "b(X) :- something(X).");
+		testLaxGrounderHeuristicTolerance(program, 0, literal("a", "X"), 1, 2, 0, true, Arrays.asList(2));
+	}
+
+	@Test
+	public void testLaxGrounderHeuristicTolerance_1_accept_two_substitutions() {
+		Program program = PARSER.parse("a(1). "
+				+ "c(X) :- a(X), b(X,Y). "
+				+ "b(X,Y) :- something(X,Y).");
+		testLaxGrounderHeuristicTolerance(program, 0, literal("a", "X"), 1, 1, new ThriceTruth[] {TRUE, TRUE}, 2, true, Arrays.asList(0, 0));
+	}
+
+	@Test
+	public void testLaxGrounderHeuristicTolerance_1_accept_accept_two_substitutions_with_different_remaining_tolerances() {
+		Program program = PARSER.parse("a(1). "
+				+ "c(X) :- a(1), b(X,Y). "
+				+ "b(X,Y) :- something(X,Y).");
+		testLaxGrounderHeuristicTolerance(program, 0, literal("a", 1), 1, 1, new ThriceTruth[] {null, null}, 2, true, Arrays.asList(1, 1));
+	}
+
+	@Test
+	public void testLaxGrounderHeuristicTolerance_2_reject() {
+		Program program = PARSER.parse("a(1). "
+				+ "c(X) :- a(X), b(X), b(X+1), b(X+2). "
+				+ "b(X) :- something(X).");
+		testLaxGrounderHeuristicTolerance(program, 0, literal("a", "X"), 1, 2, 0, false, Arrays.asList(3));
+	}
+
+	@Test
+	public void testLaxGrounderHeuristicTolerance_2_accept_multiple_facts_of_same_variable() {
+		Program program = PARSER.parse("a(1). b(1). "
+				+ "c(X) :- a(X), b(X), b(X+1), b(X+2). "
+				+ "b(X) :- something(X).");
+		testLaxGrounderHeuristicTolerance(program, 0, literal("a", "X"), 1, 2, 0, true, Arrays.asList(2));
+	}
+
+	private void testLaxGrounderHeuristicTolerance(Program program, int ruleID, Literal startingLiteral, int startingInstance, int tolerance, int nTrueBs, boolean expectNoGoods, List<Integer> expectedNumbersOfUnassignedPositiveBodyAtoms) {
+		ThriceTruth[] truthsOfB = new ThriceTruth[nTrueBs];
+		for (int i = 0; i < nTrueBs; i++) {
+			truthsOfB[i] = ThriceTruth.TRUE;
+		}
+		testLaxGrounderHeuristicTolerance(program, ruleID, startingLiteral, startingInstance, tolerance, truthsOfB, 1, expectNoGoods, expectedNumbersOfUnassignedPositiveBodyAtoms);
+	}
+
+	private void testLaxGrounderHeuristicTolerance(Program program, int ruleID, Literal startingLiteral, int startingInstance, int tolerance, ThriceTruth[] truthsOfB, int arityOfB, boolean expectNoGoods, List<Integer> expectedNumbersOfUnassignedPositiveBodyAtoms) {
```

Comment:
It would be good if those private helpers were a bit documented.

Reply:
Done.